### PR TITLE
Add `ThrowAsync`, `ThrowExactlyAsync` and `NotThrowAsync`

### LINKED ISSUE #851: Should throw API question/proposition
### Description

Currently the usage of the `Should.Throw` construction in the context of the async method looks like this:
```csharp
public void SomeTestMethod()
{
   Func<Task> f = async () =>
   {
          await Fail();
   };

  f.Should().Throw<SomeException>();
}
```

What if we had more idiomatic(in terms of the C# language) API of the `Should().Throw()` construction:
```csharp
public async Task SomeTestMethod()
{
   Func<Task> f = async () =>
   {
          await Fail();
   };

  await f.Should().Throw<SomeException>();
}
```

**Pros:**

1. Performance 
async/await all the way - in the context of a usual unit test the benefit might be not visible, since most of the mocked dependencies will be in-memory, there will be no io, hence no boost at all.
On the other hand, we use this library on a slightly broader scope - we do have tests, which might have an actual db write/read and the xunit runner, for instance, which runs lots of tests like this in parallel, could benefit from this by not holding a running thread.

1. Easier to implement - async all the way, no .Result or .Wait(), no tricks to overcome potential deadlock when blocking on the async pipeline.

**Cons:**
It seems like a breaking change

### PR DESCRIPTION
**Pull request by krajek**
Closes #851.

### CODE REVIEW

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 1]
```diff
@@ -125,28 +185,61 @@ private static Exception GetFirstNonAggregateException(Exception exception)
...
+            {
+                return InterceptException(exception);
+            }
+
+            return null;
+        }
+
+        private Exception InterceptException(Exception exception)
+        {
+            var ar = exception as AggregateException;
```
[the remark this answers is not in the document]
Comment (krajek) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: Not on this one. It's not only downcast but also reading `InnerException`. I think that this version is more readable.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 2]
```diff
@@ -101,16 +123,54 @@ public void NotThrow<TException>(string because = "", params object[] becauseArg
...
+            {
+                await Task.Run(Subject);
+            }
+            catch (Exception exception)
+            {
+                NotThrow<TException>(because, becauseArgs, exception);
+            }
+        }
+
+        private static void NotThrow(string because, object[] becauseArgs, Exception exception)
```
[the remark this answers is not in the document]
Comment (krajek) [reply Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: Appplied, thanks.

**Comment by krajek** [comment]
@jnyrup I have just added `ThrowExactlyAsync`. Ready to merge from my side.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 3]
```diff
@@ -101,16 +123,54 @@ public void NotThrow<TException>(string because = "", params object[] becauseArg
...
+        {
+            Exception nonAggregateException = GetFirstNonAggregateException(exception);
+
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .FailWith("Did not expect any exception{reason}, but found a {0} with message {1}.",
+                    nonAggregateException.GetType(), nonAggregateException.ToString());
+        }
+
+        private static void NotThrow<TException>(string because, object[] becauseArgs, Exception exception) where TException : Exception
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: 🔧 also make the `Exception` parameter the first here.

File: Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs [hunk 4]
```diff
@@ -125,28 +185,61 @@ private static Exception GetFirstNonAggregateException(Exception exception)
             return nonAggregateException;
         }
 
-        private Exception InvokeSubjectWithInterception()
+        private ExceptionAssertions<TException> Throw<TException>(string because, object[] becauseArgs, Exception exception)
```
Comment (jnyrup) [inline Src/FluentAssertions/Specialized/AsyncFunctionAssertions.cs]: 🔧 also make the `Exception` parameter the first here.

**Comment by nerdpad** [comment]
Any plans of release this soon? I just started using FluentAssertion and this is exactly what I needed.

**Comment by jnyrup** [comment]
@nerdpad yes, we're also quite eager to get out a new release with all the new features, fixes and other improvements. 
We're finishing up the last PRs for 5.5.0, see the remaining ones here https://github.com/fluentassertions/fluentassertions/milestone/12

**Pull request merged**